boskos/cleaner/v2/cleaner.go

```diff
@@ -0,0 +1,125 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v2
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"k8s.io/test-infra/boskos/cleaner"
+	"k8s.io/test-infra/boskos/common"
+	"k8s.io/test-infra/boskos/crds"
+)
+
+const controllerName = "boskos-cleaner"
+
+// Add creates a new claner controller
+func Add(mgr manager.Manager, boskosClient cleaner.RecycleBoskosClient, namespace string) error {
+	reconciler := &reconciler{
+		ctx:          context.Background(),
+		client:       mgr.GetClient(),
+		boskosClient: boskosClient,
+		namespace:    namespace,
+	}
+
+	c, err := controller.New(controllerName, mgr, controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: 4,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %v", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &crds.ResourceObject{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to create watch: %v", err)
+	}
+
+	return nil
+}
+
+type reconciler struct {
+	ctx          context.Context
+	client       ctrlruntimeclient.Client
+	boskosClient cleaner.RecycleBoskosClient
+	recycleFunc  func(cleaner.RecycleBoskosClient, *common.Resource)
+	namespace    string
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log := logrus.WithField("resource-name", request.Name)
+	err := r.reconcile(log, request)
+	if err != nil {
+		log.WithError(err).Error("Reconciliation error")
+	}
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) reconcile(log *logrus.Entry, request reconcile.Request) error {
+	resourceObject := &crds.ResourceObject{}
+	if err := r.client.Get(r.ctx, request.NamespacedName, resourceObject); err != nil {
+		if kerrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get object %s: %v", request.NamespacedName.String(), err)
+	}
+
+	// We only care about unowned resources in ToBeDeleted state
+	if (resourceObject.Status.Owner != "" && resourceObject.Status.Owner != controllerName) || resourceObject.Status.State != common.ToBeDeleted {
+		return nil
+	}
+
+	isDynamic, err := r.isResourceDynamic(resourceObject)
+	if err != nil {
+		return fmt.Errorf("failed to check if resource is dynamic: %v", err)
+	}
+	if !isDynamic {
+		return nil
+	}
+
+	resourceObject.Status.Owner = controllerName
+	if err := r.client.Update(r.ctx, resourceObject); err != nil {
```

Comment:
Is this a common pattern? Why not let the controller race with itself if it gets multiple events to handle and let one win?

Reply:
The controller doesn't race with itself because it has leader election and the underlying machinery makes sure that only one routine works on a given key.

When using multiple controller, just trying optimistically and being okay with sometimes failing the update would be the correct approach, this here is done only to prevent boskos from somehow leasing out the resource.

Reply:
That would only occur if someone actually tried to lease a toBeDeleted resource, right? I just wonder if we need to double the API calls with this approach or if we can just do the state update and let other clients retry if their updates get muddled.

Reply:
Its more like:

* We find the empty resource and start doing stuff
* Someone else acquires the resource, boskos updates the owner field
* Our state transition fails, because the object got changed

Hypothetically this may be okay because it will eventually get back to unowned, we clean again and transition the state. But it muddles with the boskos concept of "handing out crds to arbitrary actors that do arbitrary things with it".